Bug#80 Too many sync tasks on UI thread for MatchActivity

limphilip opened the pull request:
#80 

This pull request is to solve the synchronicity issue for the `MatchActivity `causing it to crash, or to be extremely slow. The following changes were made:

- In the `MatchActivity`, the `OnCreate()` method loads the view, and captures the elements of importance in the view, but does not perform synchronously the request sent to the server to obtain the list of users. That way the UI Thread remains active while a new class name `FindMatchAsyncTask` is created under a new package named `Tasks`. Once this task is finished, a callback function found in the `MatchActivity` is invoked. The presence and implementation of this method is enforced by the interface `IFindMatchCompleteCallback`.
- A toast is added when declining the request of a user.
- When looking for match, the match activity will hide the buttons that depend on the results of the async task. When the results of the tasks are received, the buttons re-appear. Texts dependent of the results are also replaced with "Loading..." while waiting for the response. 
- Reverse geocoding now uses `Android.Location` for asynchronicity and compatibility with Android phones.

The following **async methods** are now available from the mapper & gateway:
- `GetUserFacebookProfilePictureAsync()`: This converts the image to a bitmap asynchronously.

Sometimes, async methods are not sufficient, or may require callback methods. It may also be cleaner and more readable to separate the task performed in the background, and abstract that logic from the activities. The following **async tasks** are now available, each with their callback interface:
 - `GetReviewsAverageAsyncTask`: Obtain all the reviews for a given user, and compute the average asynchronously.
- `FindMatchAsyncTask`: Obtains all the matched requests asynchronously.

Note on async tasks: 
Activities have visibility over the Tasks, and the Tasks have visibility over the Mappers.

limphilip committed:
Merge remote-tracking branch 'refs/remotes/origin/dev' into bug#80/TooManySyncTasksUIThreads

limphilip committed:
#80 Initial implementation of FindMatchAsyncTask

Finding a match was the most expensive operation in term of time
consumption when loading the match activity. It was also performed on
the UI Thread, which blocked the whole main thread of the application,
causing it to crash. This implementation allows the call to obtain the
matches from the server to be done asynchronously while keeping the UI
thread running. Once done, the async task calls a call back function in
the MatchActivity, which now implements the interface
IOnFindMatchCompleteCallback. The loading of the match is still slow,
but the application does not crash. A loading screen may be required
while waiting for the matches to be found.

limphilip committed:
#80 Updated the interface of the FindMatchAsyncTask to take a request instead of data to build the request

The MatchActivity needs to have access to the ID of the user's request
in order to send the confirmation to the server if the user accepts the
match. Thus, the FindMatchAsyncTask cannot build it, or if it did, it
would have to return the list of request as well as the request of the
user itself. Thus, the creation of the user request is made in the
MatchActivity rather than in the AsyncTask.

limphilip committed:
Merge remote-tracking branch 'refs/remotes/origin/dev' into bug#80/TooManySyncTasksUIThreads

# Conflicts:
#	mRides-app/Activities/MatchActivity.cs

limphilip committed:
#80 Post-merge fix in MatchActivity

The merge added all the element listener of the activities in the method
UpdateDisplay(), while they should be in OnCreate(). They have been
moved in the appropriate place.

limphilip committed:
#80 RunOnUiThread when calling UpdateDisplay in MatchActivity

The UpdateDisplay updates elements on the match activity screen and thus
must be called on the UI thread.

limphilip committed:
#80 Hide clickable items while fetching asynchronously

Hide the clickable items from the user that causes a null pointer
exception if clicked before the response from the asynchronous request
has been received and processed. Added a "loading" string while waiting
for the response from the async call.

limphilip committed:
#80 Forgot to include the most important file for the previous commit

This commit contains the updated file that will hide/set visible the
buttons that can be clicked before receiving the async response, causing
the null pointer exception (see previous commit for detail).

limphilip committed:
#80 Removed the map creation duplication from the UpdateDisplay

The map was created repetitively, instead, the same map object should be
used and new matched users should be displayed. Map creation removed
from UpdateDisplay(). Comments added for readability and analyzability.
Setting elements to visible should be performed on UI thread, moved
those action on a method called on UI thread.

limphilip committed:
Merge remote-tracking branch 'refs/remotes/origin/dev' into bug#80/TooManySyncTasksUIThreads

limphilip committed:
#80 Added toast when declining a match

Since we are now asynchronous, it may be difficult for the user to know
whether or not the action has completed successfully (Feedback). So a
toast has been added along with the string.

limphilip committed:
Merge remote-tracking branch 'refs/remotes/origin/dev' into bug#80/TooManySyncTasksUIThreads

# Conflicts:
#	mRides-app/Resources/Resource.Designer.cs

limphilip committed:
#80 The role of the user defined on creation

The role of the user never changes for a matching session. It does not
need to be reloaded for every matched user. It is thus moved to
OnCreate().

Removed the FindViewById<>() in the UpdateDisplay() since they are
already present from the OnCreate()

limphilip committed:
#80 GetReviewsAverageAsyncTask and Facebook profile picture conversion async

- UserGateway is now able to obtain the facebook profile picture and
convert the picture asynchronously
- UserMapper is updated to provide the support of asynchronous method
for obtaining Facebook profile picture
- The folder "Callbacks" has been added under Tasks
- New interface IOnGetReviewsAverageCompleteCallback is added to enforce
the method implementation of the callback method when an average of
reviews is finished computing
- FindMatchAsyncTask has been updated to have visibility over the new
Callbacks package
- GetReviewsAverageAsyncTask is added to obtain all feedbacks of a user
and compute the average asynchronously. Useful for statistics or to
display one representative number for the rating of a user like in the
match activity.

limphilip committed:
#80 Reverse geocoding is now using Android.Location

Android.Location provides a built-in async method to reverse geocode,
which makes it easier than finding a way to call asynchronously Google
API to reverse geocode. This fixes the issue of crashing on the phone.

limphilip commented:
Please do NOT close issue #80 yet after merging, we _may_ need to change other methods to be asynchronous in other activities if performance issues arise.

Ideawin commented on mRides-app/Tasks/FindMatchAsyncTask.cs:
```diff
@@ -0,0 +1,78 @@
+        private ConsoleMapper consoleMapper;
+        
+        /// <summary>
+        /// 
```
Don't forget to write a description of each method inside this class along with the parameters :) They are all blank

limphilip replied:
Good catch! I totally forgot about them. I will add them shortly.

limphilip committed:
#80 Added comments documentation for FindMatchAsyncTask

Added comments documentation for FindMatchAsyncTask for readability.
They were previously forgotten.

Ideawin commented:
Good job on this! I just tested it out and the matching activity is loading very fast compared to previously! Merging
